File: samples/RevitMcpToolSet/Utilities/OperationOutcome.cs
using RevitMcpToolSet.Data;

namespace RevitMcpToolSet.Utilities;

internal class OperationOutcome
{
    private int _successCount;
    private readonly List<ToolError> _failures = [];

    public void RecordSuccess() => _successCount++;

    public void Record(bool success, string message, long elementId)
    {
        if (success)
            _successCount++;
        else
            _failures.Add(ToolErrorHelper.FromMessage(message, elementId));
    }

    public void RecordFailure(long elementId, Exception ex)
        => _failures.Add(ToolErrorHelper.FromException(ex, elementId));

    public void RecordFailure(long elementId, string message)
        => _failures.Add(ToolErrorHelper.FromMessage(message, elementId));

    public void RecordFailure(ToolError error)
        => _failures.Add(error);

    public object Summarize()
        => new
        {
            success_count = _successCount,
            failure_count = _failures.Count,
            failures = _failures.Count > 0 ? _failures : null,
        };

    public int SuccessCount => _successCount;
    public IReadOnlyList<ToolError> Failures => _failures;
}

File: samples/PythonDemo/mcp_toolset/shared/operation_outcome.py
"""Partial-success aggregation for write tools."""

from dto.common import ToolErrorEntry


class OperationOutcome:
    def __init__(self) -> None:
        self._success_count = 0
        self._failures: list[ToolErrorEntry] = []

    def record_success(self) -> None:
        self._success_count += 1

    def record(
        self, success: bool, message: str, element_id: int | None = None
    ) -> None:
        if success:
            self._success_count += 1
        else:
            self._failures.append(ToolErrorEntry.from_message(message, element_id))

    def record_failure(self, element_id: int | None, message: str) -> None:
        self._failures.append(ToolErrorEntry.from_message(message, element_id))

    def record_exception(self, element_id: int | None, exc: Exception) -> None:
        self._failures.append(ToolErrorEntry.from_exception(exc, element_id))

    def summarize(self) -> dict:
        return {
            "success_count": self._success_count,
            "failure_count": len(self._failures),
            "failures": [f.model_dump(by_alias=True) for f in self._failures] or None,
        }
